presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -261,18 +262,39 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
                     .collect(toImmutableList());
 
             List<List<Symbol>> inputs = new ArrayList<>();
+            List<Map<Symbol, Symbol>> inputToOutputs = new LinkedList<>();
             for (int i = 0; i < node.getInputs().size(); i++) {
                 inputs.add(new ArrayList<>());
+                inputToOutputs.add(new HashMap<>());
             }
             Set<Symbol> addedOutputs = new HashSet<>();
             ImmutableList.Builder<Symbol> outputs = ImmutableList.builder();
             for (int symbolIndex = 0; symbolIndex < node.getOutputSymbols().size(); symbolIndex++) {
                 Symbol canonicalOutput = canonicalize(node.getOutputSymbols().get(symbolIndex));
-                if (addedOutputs.add(canonicalOutput)) {
+
+                List<Symbol> outputAliases = new ArrayList<>();
+                for (int i = 0; i < node.getInputs().size(); i++) {
+                    List<Symbol> input = node.getInputs().get(i);
+                    Symbol canonicalInput = canonicalize(input.get(symbolIndex));
+                    if (inputToOutputs.get(i).containsKey(canonicalInput)) {
+                        outputAliases.add(inputToOutputs.get(i).get(canonicalInput));
+                    }
+                }
+
+                if (!addedOutputs.contains(canonicalOutput)) {
+                    if (outputAliases.size() == node.getInputs().size() && !outputAliases.isEmpty()
+                            && outputAliases.stream().allMatch(symbol -> symbol.equals(outputAliases.get(0)))) {
+                        map(canonicalOutput, outputAliases.get(0));
+                        continue;
+                    }
+
+                    addedOutputs.add(canonicalOutput);
```

Comment:
please try to extract this optimization as separate one. If these two kind of optimization are mixed together it is hard to follow what is happening here.